frontend/middleware_validatesubscription.go

```diff
@@ -0,0 +1,83 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+	"github.com/Azure/ARO-HCP/internal/api/subscription"
+)
+
+const (
+	UnregisteredSubscriptionStateMessage = "The subscription %s is not registered for this provider %s. Please re-register the subscription."
+	SubscriptionMissingMessage           = "The request is missing required parameter '%s'."
+	InvalidSubscriptionStateMessage      = "The subscription %s is in %s state, please ensure subscription is eligible to use the provider %s."
+	DeletedSubscriptionMessage           = "The subscription %s is deleted and cannot be used to interact with %s."
+)
+
+type SubscriptionStateMuxValidator struct {
+	cache *Cache
+}
+
+func NewSubscriptionStateMuxValidator(c *Cache) *SubscriptionStateMuxValidator {
+	return &SubscriptionStateMuxValidator{
+		cache: c,
+	}
+}
+
+// MiddlewareValidateSubscriptionState validates the state of the subscription as outlined by
+// https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/subscription-lifecycle-api-reference.md
+func (s *SubscriptionStateMuxValidator) MiddlewareValidateSubscriptionState(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	subscriptionId := r.PathValue(strings.ToLower(PathSegmentSubscriptionID))
+	sub, exists := s.cache.GetSubscription(subscriptionId)
+	if subscriptionId == "" {
```

Comment:
Shouldn't we check the cache entry exists first before checking the contents of the string?

Reply:
I think you mean, we should check if the string is empty, before checking if the string is key for our cache. I'll fix this.

Reply:
Ah yeah I misread.  

- Check string empty
- fetch from cache
- check if cache miss
- ...

Reply:
this is now done.